### PR TITLE
InputStream implementations should honor contract.

### LINKED ISSUE #932: MessageFormatInputStream does not conform with InputStream
"If len is zero, then no bytes are read and 0 is returned; otherwise, there is an attempt to read at least one byte. If no byte is available because the stream is at end of file, the value -1 is returned; otherwise, at least one byte is read and stored into b."  from `public int read(byte[] b, int off, int len) throws IOException` at https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html

`MessageFormatInputStream`'s `int read(byte[] b, int off, int len)` does not conform to this description, as it never returns -1.

### PR DESCRIPTION
**Pull request by lightningrob**
Fix #932

### CODE REVIEW

File: ambry-utils/src/main/java/com.github.ambry.utils/ZeroBytesInputStream.java [hunk 1]
```diff
@@ -48,6 +48,12 @@ public int read() throws IOException {
 
   @Override
   public int read(byte[] bytes, int offset, int length) throws IOException {
+    if (length == 0) {
```
Comment (vgkholla) [inline ambry-utils/src/main/java/com.github.ambry.utils/ZeroBytesInputStream.java]: this class doesn't have tests?

File: ambry-utils/src/main/java/com.github.ambry.utils/ByteBufferInputStream.java [hunk 2]
```diff
@@ -67,6 +67,9 @@ public int read() throws IOException {
 
   @Override
   public int read(byte[] bytes, int offset, int length) throws IOException {
+    if (length == 0) {
```
Comment (vgkholla) [inline ambry-utils/src/main/java/com.github.ambry.utils/ByteBufferInputStream.java]: this class doesn't have tests?

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatInputStream.java [hunk 3]
```diff
@@ -95,7 +95,7 @@ public int read(byte b[], int off, int len) throws IOException {
         totalRead += bytesToRead;
       }
     }
-    return totalRead;
+    return totalRead > 0 ? totalRead : -1;
```
Comment (vgkholla) [inline ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatInputStream.java]: this class doesn't have tests?

**Comment by pnarayanan** [comment]
Line 57 of `MessageFormatInputStream.java` has a bug because it does an implicit cast of a `byte` (return type of `ByteBuffer#get()`) to an `int` (return type of `InputStream#read()`). `InputStream` requires that the `read()` return the next `byte` as an `int` in the range 0 - 255 (basically return the next `byte` in the least significant `byte` of the `int` while keeping the higher bits 0). With negative values the implicit cast can lead to unwanted consequences, which is why you see code elsewhere that explicitly clears out the significant bits (by doing `&0xFF`).

To illustrate, here is what happens with -1:

```
byte:                                                     1111 1111
int (implicit cast):        1111 1111 1111 1111 1111 1111 1111 1111
int (byte & 0xFF):          0000 0000 0000 0000 0000 0000 1111 1111
```

The value returned will be incorrect and in fact a -1 int return is treated as EOF for `InputStream`. I suspect the single byte `read()` method does not get called in production or test code which is why this was never caught (we end up calling the other `read()` method always). Would be a good idea to have a test for the former.

File: ambry-utils/src/main/java/com.github.ambry.utils/ZeroBytesInputStream.java [hunk 4]
```diff
@@ -48,6 +48,12 @@ public int read() throws IOException {
...
+      return 0;
+    }
```
Comment (pnarayanan) [inline ambry-utils/src/main/java/com.github.ambry.utils/ZeroBytesInputStream.java]: there are a few checks that the `InputStream#read()` is supposed to do as per the docs. Most implementations that override this method copy that verbatim:
```java
        if (b == null) {
            throw new NullPointerException();
        } else if (off < 0 || len < 0 || len > b.length - off) {
            throw new IndexOutOfBoundsException();
        } else if (len == 0) {
            return 0;
        }
```
I checked `ByteBufferInputStream#read()` and for that the bounds check happens anyway within `ByteBuffer#get()`, so this check is unnecessary. But for this class, it is necessary.

File: ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatInputStream.java (hunk 3 above)
Comment (lightningrob) [reply ambry-messageformat/src/main/java/com.github.ambry.messageformat/MessageFormatInputStream.java]: It does, and I enhanced it.

File: ambry-utils/src/main/java/com.github.ambry.utils/ByteBufferInputStream.java (hunk 2 above)
Comment (lightningrob) [reply ambry-utils/src/main/java/com.github.ambry.utils/ByteBufferInputStream.java]: It does, will add a test case.

File: ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java [hunk 5]
```diff
@@ -158,6 +158,40 @@ public static void assertInputStreamEqual(InputStream actual, InputStream expect
...
+   */
+  public static void readInputStreamAndValidateSize(InputStream inputStream, long expectedLength) throws IOException {
+    int readVal = 0;
+    long numRead = 0;
+    do {
+      readVal = inputStream.read();
+      numRead++;
+    } while (readVal != -1);
+    numRead--;
+    Assert.assertEquals(expectedLength, numRead);
```
Comment (vgkholla) [inline ambry-utils/src/test/java/com.github.ambry.utils/TestUtils.java]: message?

**Pull request merged**